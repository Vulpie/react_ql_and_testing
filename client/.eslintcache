[{"/home/vulpie/Desktop/projects/ql/client/src/index.js":"1","/home/vulpie/Desktop/projects/ql/client/src/App.js":"2","/home/vulpie/Desktop/projects/ql/client/src/Components/Home.js":"3","/home/vulpie/Desktop/projects/ql/client/src/Components/Lists/BookList.js":"4","/home/vulpie/Desktop/projects/ql/client/src/Components/Lists/AuthorList.js":"5","/home/vulpie/Desktop/projects/ql/client/src/Components/Book/Book.js":"6","/home/vulpie/Desktop/projects/ql/client/src/Components/Author/Author.js":"7","/home/vulpie/Desktop/projects/ql/client/src/Components/Author/AddAuthor.js":"8","/home/vulpie/Desktop/projects/ql/client/src/Components/Partials/LinkButton.js":"9","/home/vulpie/Desktop/projects/ql/client/src/Components/Partials/HomeButton.js":"10","/home/vulpie/Desktop/projects/ql/client/src/Components/Book/AddBook.js":"11"},{"size":209,"mtime":1610839960276,"results":"12","hashOfConfig":"13"},{"size":1107,"mtime":1611438731037,"results":"14","hashOfConfig":"13"},{"size":315,"mtime":1610894367685,"results":"15","hashOfConfig":"13"},{"size":644,"mtime":1611421730355,"results":"16","hashOfConfig":"13"},{"size":613,"mtime":1611402966557,"results":"17","hashOfConfig":"13"},{"size":1147,"mtime":1611430961232,"results":"18","hashOfConfig":"13"},{"size":243,"mtime":1610889405967,"results":"19","hashOfConfig":"13"},{"size":1039,"mtime":1610878985504,"results":"20","hashOfConfig":"13"},{"size":823,"mtime":1611207371678,"results":"21","hashOfConfig":"13"},{"size":303,"mtime":1611354760355,"results":"22","hashOfConfig":"13"},{"size":1447,"mtime":1611438891194,"results":"23","hashOfConfig":"13"},{"filePath":"24","messages":"25","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0,"usedDeprecatedRules":"26"},"1k8yqt7",{"filePath":"27","messages":"28","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0},{"filePath":"29","messages":"30","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0,"usedDeprecatedRules":"26"},{"filePath":"31","messages":"32","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0,"usedDeprecatedRules":"26"},{"filePath":"33","messages":"34","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0,"usedDeprecatedRules":"26"},{"filePath":"35","messages":"36","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0},{"filePath":"37","messages":"38","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0,"usedDeprecatedRules":"26"},{"filePath":"39","messages":"40","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0,"usedDeprecatedRules":"26"},{"filePath":"41","messages":"42","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0,"usedDeprecatedRules":"26"},{"filePath":"43","messages":"44","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0,"usedDeprecatedRules":"26"},{"filePath":"45","messages":"46","errorCount":0,"warningCount":0,"fixableErrorCount":0,"fixableWarningCount":0},"/home/vulpie/Desktop/projects/ql/client/src/index.js",[],["47","48"],"/home/vulpie/Desktop/projects/ql/client/src/App.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Home.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Lists/BookList.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Lists/AuthorList.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Book/Book.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Author/Author.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Author/AddAuthor.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Partials/LinkButton.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Partials/HomeButton.js",[],"/home/vulpie/Desktop/projects/ql/client/src/Components/Book/AddBook.js",[],{"ruleId":"49","replacedBy":"50"},{"ruleId":"51","replacedBy":"52"},"no-native-reassign",["53"],"no-negated-in-lhs",["54"],"no-global-assign","no-unsafe-negation"]