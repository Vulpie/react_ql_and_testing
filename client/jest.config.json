{
	"name": "my-project",
	"jest": {
		"verbose": true
	},
	"setupFilesAfterEnv": ["raf/polyfill", "<rootDir>/src/tests/setupTests.js"],
	"testPathIgnorePatterns": ["<rootDir>/node_modules/"],
	"moduleNameMapper": {
		"\\.(jpg|ico|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/src/tests/__mocks__/fileMock.js",
		"\\.(css|less)$": "<rootDir>/src/tests/__mocks__/fileMock.js"
	}
}
