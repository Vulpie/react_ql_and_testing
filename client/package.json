{
	"name": "client",
	"version": "0.1.0",
	"private": true,
	"proxy": "http://localhost:5000",
	"babel": {
		"presets": [
			"@babel/env",
			"@babel/react"
		]
	},
	"dependencies": {
		"@apollo/client": "^3.3.7",
		"@testing-library/jest-dom": "^5.11.9",
		"@testing-library/react": "^11.2.3",
		"@testing-library/user-event": "^12.6.0",
		"babel-jest": "^26.6.3",
		"enzyme": "^3.11.0",
		"enzyme-adapter-react-16": "^1.15.6",
		"enzyme-to-json": "^3.6.1",
		"graphql": "^15.4.0",
		"jest-dom": "^4.0.0",
		"react": "^17.0.1",
		"react-dom": "^17.0.1",
		"react-router-dom": "^5.2.0",
		"react-routes": "^0.2.6",
		"react-scripts": "4.0.1",
		"react-test-renderer": "^17.0.1",
		"web-vitals": "^0.2.4"
	},
	"scripts": {
		"start": "react-scripts start",
		"build": "react-scripts build",
		"test": "jest --config=jest.config.json --watch ",
		"eject": "react-scripts eject"
	},
	"eslintConfig": {
		"extends": [
			"react-app",
			"react-app/jest"
		]
	},
	"browserslist": {
		"production": [
			">0.2%",
			"not dead",
			"not op_mini all"
		],
		"development": [
			"last 1 chrome version",
			"last 1 firefox version",
			"last 1 safari version"
		]
	},
	"devDependencies": {
		"@wojtekmaj/enzyme-adapter-react-17": "^0.4.1",
		"identity-obj-proxy": "^3.0.0"
	}
}
