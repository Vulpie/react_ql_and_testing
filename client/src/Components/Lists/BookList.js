import React from 'react'
import { gql, useQuery } from '@apollo/client'
import HomeButton from '../Partials/HomeButton'
import { Link } from 'react-router-dom'

export const GET_BOOKS = gql`
	{
		books {
			id
			name
			genre
			author {
				name
			}
		}
	}
`

export default function BookList() {
	const { loading, error, data } = useQuery(GET_BOOKS)
	if (loading) return <p>Loading list of books ...</p>
	if (error) return <p>Error -.-</p>

	return (
		<div className='component'>
			<HomeButton />
			<div className='grid-container'>
				{data.books.map(({ id, name, genre, author }, index) => (
					<Link to={`/book/details/${id}`} key={'book_' + index}>
						<div className='grid-container__block'>
							<div className='grid-container__block_property'>
								<p className='grid-container__block_property_header'>
									Title:
								</p>
								<p className='grid-container__block_property_value'>
									{name}
								</p>
							</div>
							<div className='grid-container__block_property'>
								<p className='grid-container__block_property_header'>
									Genre:
								</p>
								<p className='grid-container__block_property_value'>
									{genre}
								</p>
							</div>
							<div className='grid-container__block_property'>
								<p className='grid-container__block_property_header'>
									Author:
								</p>
								<p className='grid-container__block_property_value'>
									{author.name}
								</p>
							</div>
						</div>
					</Link>
				))}
			</div>
		</div>
	)
}
