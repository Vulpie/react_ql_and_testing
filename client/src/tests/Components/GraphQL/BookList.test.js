import React from 'react'
import { MockedProvider } from '@apollo/client/testing'
import { act, render } from '@testing-library/react'
import BookList, { GET_BOOKS } from '../../../Components/Lists/BookList'
import { BrowserRouter as Router } from 'react-router-dom'

const { GraphQLError } = require('graphql')

describe('Render BookList component', () => {
	test('Loading state', () => {
		const mocks = []
		const { getByText, container } = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<Router>
					<BookList />
				</Router>
			</MockedProvider>
		)

		expect(getByText('Loading list of books ...')).toBeInTheDocument()
		expect(container).toMatchSnapshot()
	})
	test('Sucess state', async () => {
		const mocks = [
			{
				request: { query: GET_BOOKS },
				result: {
					data: {
						books: [
							{
								name: 'Book1',
								genre: 'Fantasy',
								author: {
									name: 'Author1',
								},
							},
						],
					},
				},
			},
		]
		const { getByText, container } = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<Router>
					<BookList />
				</Router>
			</MockedProvider>
		)
		await act(async () => {
			await new Promise((resolve) => setTimeout(resolve, 0))
		})
		expect(getByText('Title: Book1')).toBeInTheDocument()
		expect(getByText('Genre: Fantasy')).toBeInTheDocument()
		expect(getByText('Author: Author1')).toBeInTheDocument()
		expect(container).toMatchSnapshot()
	})
	test('Network error', async () => {
		const mocks = [
			{
				request: { query: GET_BOOKS },
				error: new Error('An error occurred'),
			},
		]
		const { getByText, container } = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<Router>
					<BookList />
				</Router>
			</MockedProvider>
		)

		await act(async () => {
			await new Promise((resolve) => setTimeout(resolve, 0))
		})

		expect(getByText('Error -.-')).toBeInTheDocument()
		expect(container).toMatchSnapshot()
	})
	test('GraphQL error', async () => {
		const mocks = [
			{
				request: { query: GET_BOOKS },
				result: {
					errors: [new GraphQLError('Error!')],
				},
			},
		]
		const { getByText, container } = render(
			<MockedProvider mocks={mocks} addTypename={false}>
				<Router>
					<BookList />
				</Router>
			</MockedProvider>
		)

		await act(async () => {
			await new Promise((resolve) => setTimeout(resolve, 0))
		})

		expect(getByText('Error -.-')).toBeInTheDocument()
		expect(container).toMatchSnapshot()
	})
})
