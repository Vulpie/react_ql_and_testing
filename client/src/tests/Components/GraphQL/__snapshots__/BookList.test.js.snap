// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render BookList component GraphQL error 1`] = `
<div>
  <p>
    Error -.-
  </p>
</div>
`;

exports[`Render BookList component Loading state 1`] = `
<div>
  <p>
    Loading list of books ...
  </p>
</div>
`;

exports[`Render BookList component Network error 1`] = `
<div>
  <p>
    Error -.-
  </p>
</div>
`;

exports[`Render BookList component Sucess state 1`] = `
<div>
  <div>
    <a
      href="/"
    >
      <img
        alt="Return button"
        class="home-button"
        src=""
      />
    </a>
    <div>
      <p>
        Title: 
        Book1
      </p>
      <p>
        Genre: 
        Fantasy
      </p>
      <p>
        Author: 
        Author1
      </p>
    </div>
  </div>
</div>
`;
